src/components/csv-download-button.tsx

```diff
@@ -0,0 +1,59 @@
+'use client';
+
+import { useFilterState } from '@/app/common/hooks/use-filter-state';
+import { CSVDownloadObjectType, getTxnData } from '@/common/utils/csv-download-utils';
+import { Box, Icon } from '@/ui/components';
+import { CSVLink } from 'react-csv';
+import { Caption } from '@/ui/typography';
+import React, { memo, useRef, useState } from 'react';
+import { FiDownload } from 'react-icons/fi';
+import { useQueryClient } from 'react-query';
+
+import { useRouter } from 'next/router';
+
+export const CSVDownloadButton = memo(() => {
+  const [transactionData, setTransactionData] = useState<CSVDownloadObjectType[]>([]);
+  const queryClient = useQueryClient();
+  const { activeFilters } = useFilterState();
+  const {
+    query: { principal },
+  } = useRouter();
+  const csvLink = useRef<CSVLink & HTMLAnchorElement & { link: HTMLAnchorElement }>(null);
+
+  const donwloadCSV = () => {
+    const formattedTxnData = getTxnData(queryClient, activeFilters, principal as string);
+
+    setTransactionData(formattedTxnData);
+    console.log(formattedTxnData[0]);
+
+    // This is added to prevent immediate invocation of click function.
+    // Click should get invoked after the setState render is finished
+    setTimeout(() => {
```

Comment:
We usually avoid this approach since it can lead to race conditions and make debugging difficult. Is there any chance we could explore using a ready flag or restructuring the code to remove the need for a setTimeout and passing a ref? That would make things much smoother and more efficient in the long run. Thanks!

Reply:
Yes, I can add another state which will act as a ready flag and I can use useEffect to trigger the click event and change the ready flag to false after the click. Will that be ok?

Reply:
The above solution was not working. 
The reason for that is `csvLink.current.link.click()` This triggers the click before the state change is completed and thus CSVLink component gets empty data.

I have found two other solutions for this.
1. CSVDownload component - Flow is working fine with this one but the names of the files getting downloaded are like `28ffd53d-060c-43f4-ab16-fa921596cb2e.csv`.

2. Or, I can use two useEffects so that component gets fully rendered  before triggering the click effect.
![image](https://github.com/hirosystems/explorer/assets/20265606/220ea42a-37a7-4ffd-ad4e-9f31bad6ca91).

Please suggest which would be better or if there is some possible solution.